### PR TITLE
[SPARK-5754] [yarn] Spark AM not launching on Windows

### PR DESCRIPTION
**Pull request by cbvoxel**
This is my suggestion from the discussion in the ticket [SPARK-5745], the PR #7829 and my ideas for fixes for the situation.

I added tests, which **fail**!! I need your support to define the correct behavior.

### CODE REVIEW

**Comment by cbvoxel** [comment]
Did someone have time to look over this?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 1]
```diff
@@ -316,7 +316,9 @@ private[spark] class Client(
         destName: Option[String] = None,
         targetDir: Option[String] = None,
         appMasterOnly: Boolean = false): (Boolean, String) = {
+      logDebug("Current Path=" + path)
       val localURI = new URI(path.trim())
+      logDebug("Current Path as URI=" + localURI)
```
Comment (srowen) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: These should just be removed; at best use string interpolation but probably worth removing.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala [hunk 2]
```diff
@@ -146,11 +146,11 @@ class ExecutorRunnable(
     javaOpts += "-Xmx" + executorMemoryString
 
     // Set extra Java options for the executor, if defined
-    sys.props.get("spark.executor.extraJavaOptions").foreach { opts =>
-      javaOpts ++= Utils.splitCommandString(opts).map(YarnSparkHadoopUtil.escapeForShell)
-    }
+    sys.props.get("spark.executor.extraJavaOptions").map(
```
Comment (srowen) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala]: why `map` and not `foreach`?

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala [hunk 3]
```diff
@@ -219,6 +219,13 @@ object YarnSparkHadoopUtil {
     }
   }
 
+  def escapeForShell(arg: String): String = {
+    if(Utils.isWindows)
```
Comment (srowen) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: Missing a space after `if` and needs braces

File: yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala [hunk 4]
```diff
@@ -44,14 +44,23 @@ class YarnSparkHadoopUtilSuite extends SparkFunSuite with Matchers with Logging
         false
     }
 
-  if (!hasBash) {
+  val isWindows = {
+    val os = System.getProperty("os.name")
+    if (os.contains("Windows")) true else false
```
Comment (srowen) [inline yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala]: `if (x) true else false` is redundant -- just `x`

**Comment by vanzin** [comment]
You have a whole lot of style violations in your code. I'd recommend starting off at:
https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide

You don't need to duplicate the code in `CommandBuilderUtils`. You can add a `private[spark]` Scala class that exposes it, similar to this:
https://github.com/apache/spark/blob/3c0156899dc1ec1f7dfe6d7c8af47fa6dc7d00bf/core/src/main/scala/org/apache/spark/launcher/WorkerCommandBuilder.scala

Also, there's a lot of changes here. The original attempt at fixing this only touched the escaping function, can you explain why do you need to touch anything else? I don't have YARN running on Windows so I can't really test any of this.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala (hunk 1 above)
Comment (cbvoxel) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: Agreed!

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala [hunk 5]
```diff
@@ -412,7 +414,7 @@ private[spark] class Client(
     }
 
     // Distribute an archive with Hadoop and Spark configuration for the AM.
-    val (_, confLocalizedPath) = distribute(createConfArchive().getAbsolutePath(),
+    val (_, confLocalizedPath) = distribute(createConfArchive().toURI.getPath,
```
Comment (cbvoxel) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala]: This is a crititcal one, but it doesn't relate to the ticket at all. I get an Exception claiming that it cannot convert a windows path to an URI. In line 320 (where I put those obnoxious debug logs) it fails when using "getAbsolutePath()" here. The proper way would to get a URI and pass that, so that we'll never ever get any complications.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala (hunk 2 above)
Comment (cbvoxel) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala]: Sorry, I was looking for something like "SelectMany" from .Net and after not finding it, I forgot to turn it back. Very sloppy of me, my apologies....

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala [hunk 6]
```diff
@@ -231,7 +231,7 @@ class ExecutorRunnable(
         "2>", ApplicationConstants.LOG_DIR_EXPANSION_VAR + "/stderr")
 
     // TODO: it would be nicer to just make sure there are no null commands here
-    commands.map(s => if (s == null) "null" else s).toList
+    commands.map(s => if (s == null) "null" else s).map(YarnSparkHadoopUtil.escapeForShell).toList
```
Comment (cbvoxel) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala]: I wanted to treat all arguments in one batch, that's why I removed all "escapeForShell" invokations before this spot.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala (hunk 3 above)
Comment (cbvoxel) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: Yes!

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala [hunk 7]
```diff
@@ -244,6 +251,47 @@ object YarnSparkHadoopUtil {
...
+  /**
+   * Escapes a string for inclusion in a command line executed by Yarn. Yarn executes commands
+   * using a batch file (.cmd) Windows escaping is neccessary. The
+   * argument is enclosed in single quotes only if there are spaces inside the string.
+   * Also some key characters are escaped that have a meaning in Windows.
+   *
+   * @param arg A single argument.
+   * @return Argument quoted for execution via Yarn's generated shell script.
+   */
+  def escapeForShellWindows(arg: String): String = {
```
Comment (cbvoxel) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: @vanzin this is not a mere copy. It does add some functionality, and if I don't reinterate the string, I am left with no option but to copy 90% of the code. I wasn't sure about this earlier, but indeed, using CommandBuilderUtils would not be the answer. Or should we change CommandBuilderUtils to include the ' and the % case? The % case is critical, since cmd expects an environment variable.

File: yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala (hunk 4 above)
Comment (cbvoxel) [inline yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala]: Man.... did I write that? So sooorry....

**Comment by cbvoxel** [comment]
My sincere apologies for the lack of coding style and my sloppyness... this is (almost) unforgivable, and I will better myself!

Apart from that, I will look into this again, make modifications from your comments, will improve my coding style, and will give a suggestion as to what escapeForShellWindows really needs to be doing and formulate a test.

I wonder if anyone has experience with YARN and Spark on windows. If yes, please give your advice.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala (hunk 7 above)
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: It's fine if you have a reason to add more code. But, as I requested, please explain that in the change summary (if you don't know what that is, that's the very first comment up there under the title, which gets added to the git log). That way the explanation gets recorded in the change history, and someone later doesn't need to scratch their heads figuring out why these things were done this way.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala [hunk 8]
```diff
@@ -219,6 +219,15 @@ object YarnSparkHadoopUtil {
     }
   }
 
+  def escapeForShell(arg: String): String = {
+    if (Utils.isWindows) {
+      escapeForShellWindows(arg)
+    }
+    else {
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: nit: `else` goes in previous line.

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala [hunk 9]
```diff
@@ -244,6 +253,47 @@ object YarnSparkHadoopUtil {
     }
   }
 
+  /**
+   * Escapes a string for inclusion in a command line executed by Yarn. Yarn executes commands
+   * using a batch file (.cmd) Windows escaping is neccessary. The
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: spelling: necessary

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala [hunk 10]
```diff
@@ -244,6 +253,47 @@ object YarnSparkHadoopUtil {
...
+   * argument is enclosed in single quotes only if there are spaces inside the string.
+   * Also some key characters are escaped that have a meaning in Windows.
+   *
+   * @param arg A single argument.
+   * @return Argument quoted for execution via Yarn's generated shell script.
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: nit: YARN

File: yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala [hunk 11]
```diff
@@ -244,6 +253,47 @@ object YarnSparkHadoopUtil {
...
+    if (!needsQuotes) {
+      return arg
+    }
+
+    val quoted: StringBuilder = new StringBuilder("\"")
+    for (i <- 0 to arg.length() - 1) {
+      arg.charAt(i) match {
+        case '"' =>
+          quoted.append("\"\"")
+        case '%' =>
```
Comment (vanzin) [inline yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala]: Is there a reason why this change shouldn't be done in `CommandBuilderUtils`?

What happens if you try to run something like `bin\spark-shell.cmd --master local --name "%abcde"`?

File: yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala [hunk 12]
```diff
@@ -44,14 +44,23 @@ class YarnSparkHadoopUtilSuite extends SparkFunSuite with Matchers with Logging
         false
     }
 
-  if (!hasBash) {
+  val isWindows = {
```
Comment (vanzin) [inline yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala]: There's already `Utils.isWindows`.

File: yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala [hunk 13]
```diff
@@ -44,14 +44,23 @@ class YarnSparkHadoopUtilSuite extends SparkFunSuite with Matchers with Logging
...
+    val os = System.getProperty("os.name")
+    os.contains("Windows")
+  }
+
+  if (!hasBash && !isWindows) {
```
Comment (vanzin) [inline yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala]: So, basically:

    if (Utils.isWindows) {

    } else if (!hasBash) {

    }

File: yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala [hunk 14]
```diff
@@ -70,6 +79,31 @@ class YarnSparkHadoopUtilSuite extends SparkFunSuite with Matchers with Logging
...
+      val argLine = args.map(a => YarnSparkHadoopUtil.escapeForShell(a)).mkString(" ")
+      Files.write(("cmd /C \"@echo " + argLine + "\"").getBytes(), scriptFile)
+      scriptFile.setExecutable(true)
+System.out.println(s"Escaped Arguments: $argLine")
+      val proc = Runtime.getRuntime().exec(Array(scriptFile.getAbsolutePath()))
+      val out = new String(ByteStreams.toByteArray(proc.getInputStream())).trim()
+      val err = new String(ByteStreams.toByteArray(proc.getErrorStream()))
+      val exitCode = proc.waitFor()
+      exitCode should be (0)
+      System.out.println(s"Output: $out")
```
Comment (vanzin) [inline yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala]: Please don't print to stdout. If you really need to, extend `Logging` and use one of the log methods. But it really feels like you should just use `Utils.executeAndGetOutput`.

File: yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala [hunk 15]
```diff
@@ -70,6 +79,31 @@ class YarnSparkHadoopUtilSuite extends SparkFunSuite with Matchers with Logging
     }
   }
 
+  def cmdTest(name: String)(fn: => Unit): Unit =
+    if (isWindows) test(name)(fn) else ignore(name)(fn)
+
+  cmdTest("cmd shell script escaping") {
+    val scriptFile = File.createTempFile("script.", ".cmd", Utils.createTempDir())
```
Comment (vanzin) [inline yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala]: `Utils.createTempDir()` is unnecessary here.

**Comment by cbvoxel** [comment]
Thank you for your effort and time, it will not be in vain. I will work on this as soon as I am able and present a better solution. In the meantime, should I close this and present a new solution with all your helpful comments applied?

Thanks agian & have a good time!

**Comment by cbvoxel** [comment]
I will open up a new one with a lot less mistakes. Thanks again...!!
